pkg/frontend/querymiddleware/labels_query_optimizer.go

```diff
@@ -0,0 +1,173 @@
+	default:
+		return l.next.RoundTrip(req)
+	}
+}
+
+func (l *labelsQueryOptimizer) optimizeLabelNamesRequest(ctx context.Context, req *http.Request, parsedReq *PrometheusLabelNamesQueryRequest) (*http.Response, error) {
+	spanLog := spanlogger.FromContext(ctx, l.logger)
+
+	// Track the total number of requests processed when optimization is enabled.
+	l.totalQueries.Inc()
+
+	// Check if there are any matchers to optimize.
+	rawMatchers := parsedReq.GetLabelMatcherSets()
+	if len(rawMatchers) == 0 {
+		return l.next.RoundTrip(req)
+	}
+
+	// Optimize the matchers.
+	optimizedMatchers, optimized, err := optimizeLabelNamesRequestMatchers(rawMatchers)
+	if err != nil {
+		// Do not log an error because this error is typically caused by a malformed request, which wouldn't be actionable.
+		spanLog.DebugLog("msg", "failed to optimize label names matchers", "err", err)
+		return l.next.RoundTrip(req)
+	}
+
+	if !optimized {
+		return l.next.RoundTrip(req)
+	}
+
+	// Create a new request with optimized matchers.
+	optimizedParsedReq, err := parsedReq.WithLabelMatcherSets(optimizedMatchers)
+	if err != nil {
+		level.Error(l.logger).Log("msg", "failed to clone label names request with optimized matchers", "err", err)
+		l.failedQueries.Inc()
+		return l.next.RoundTrip(req)
+	}
+
+	// Encode the optimized request back to HTTP.
+	optimizedReq, err := l.codec.EncodeLabelsSeriesQueryRequest(ctx, optimizedParsedReq)
+	if err != nil {
+		level.Error(l.logger).Log("msg", "failed to encode label names request with optimized matchers", "err", err)
+		l.failedQueries.Inc()
+		return l.next.RoundTrip(req)
+	}
+
+	// Track successful optimization.
+	l.rewrittenQueries.Inc()
+	spanLog.DebugLog("msg", "optimized label names query", "original_matchers", strings.Join(parsedReq.LabelMatcherSets, " "), "optimized_matchers", strings.Join(optimizedParsedReq.GetLabelMatcherSets(), " "))
+	return l.next.RoundTrip(optimizedReq)
+}
+
+func optimizeLabelNamesRequestMatchers(rawMatcherSets []string) (_ []string, optimized bool, _ error) {
+	matcherSets, err := parser.ParseMetricSelectors(rawMatcherSets)
+	if err != nil {
+		return nil, false, err
+	}
+
+	optimizedRawMatcherSets := make([]string, 0, len(rawMatcherSets))
+
+	for _, matchers := range matcherSets {
+		// If an empty matcher `{}` was provided we need to preserve it because it's an invalid matcher
+		// for the labels API, but we want to get the actual error from downstream. In this case we just
+		// don't optimize any matchers set.
+		if len(matchers) == 0 {
+			return rawMatcherSets, false, nil
+		}
+
+		optimizedMatchers := make([]*labels.Matcher, 0, len(matchers))
+
+		for _, matcher := range matchers {
+			// Filter out `__name__!=""` matcher because all series in Mimir have a metric name
+			// so this matcher matches all series but very expensive to run.
+			if matcher.Name == labels.MetricName && matcher.Type == labels.MatchNotEqual && matcher.Value == "" {
```

Comment:
We could also do `foo=~".*"`.

Reply:
Good point. Will do in the follow up PR, where I add label names support too.